### ISSUE TITLE
`dead_code` warning not emitted for unused associated constant

### ISSUE THREAD

**Issue by Freax13** [rust-lang/rust#114730]
### Code

I tried this code:

```rust
struct Foo(u16);

pub fn use_foo() {
    let _ = Foo(0);
}

impl Foo {
    pub const BAR: Self = Self(123);
}

impl PartialEq<Foo> for u16 {
    fn eq(&self, other: &Foo) -> bool {
        *self == other.0
    }
}

```

I expected to see this happen: I expected to see a `dead_code` warning for `Foo::BAR`.

Instead, this happened: No warning is emitted.

### Version it worked on

It most recently worked on: Rust 1.70

### Version with regression

`rustc --version --verbose`:
```
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

### Backtrace

N/A

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

**Comment by Freax13** [comment]
searched nightlies: from nightly-2023-03-04 to nightly-2023-05-27
regressed nightly: nightly-2023-05-15
searched commit range: https://github.com/rust-lang/rust/compare/2c41369acc445d04129db40ba998dd7a89fb0d2e...18bfe5d8a9ca0e226171e98f8f4ef071790f3352
regressed commit: https://github.com/rust-lang/rust/commit/ad6ab11234ae885913229f6de2c4465bdc0d76f3

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.6</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start 1.69.0 --end 1.71.0 --regress success --access github -vvvvv -c clippy -- clippy -- -D warnings 
```
</details>

**Comment by lqd** [comment]
That's #111425, cc PR author @Bryanskiy and reviewer @petrochenkov

**Comment by Bryanskiy** [comment]
Fixed in https://github.com/rust-lang/rust/pull/112670

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23114730.20.60dead_code.60.20warning.20not.20emitted.20for.20unused.20associat.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by lqd** [comment]
Indeed, it's fixed on nightly, as well as beta. The fix will land on stable in 12 days. I'll close this but feel free to reopen in the highly unlikely chance that it appears on the next stable.

Thanks for opening an issue!
